File: seminar_Stoh/sem_b.py
import numpy as np

'''
Kod za rjesavanje b) seminarskog zadatka. Racunanje vrijenosti ⟨θ^2⟩, gdje je θ kut spina cestice u xy-ravnini,
te ln(N), gdje je N=L^2. Proucavanje ovisnosti ⟨θ^2⟩ o ln(N). Simulacija se održava na udjelu prihvacanja 40%.
'''

kT = 0.1 #fiksna temperatura za zadatak b)
beta = 1.0/kT
J = 1.0 #integral interakcije
Nk = 4 #broj Monte Carlo koraka po cestici
Nb_skip = 1500 #broj preskocenih blokova za ekvilibraciju
Nb = 10000 #broj blokova za mjerenje
acceptance = 0.4 #udio prihvacanja 40%

seed = 1208 #seed za nasumicni broj
rng = np.random.default_rng(seed)

Ls = [4, 8, 16, 32, 48, 64, 80, 96, 128] #razlicite velicine resetki
fTheta2 = open("f_theta2_vs_L.dat", "w")
fTheta2.write("# L  N  <theta^2>  ln(N)\n")

def E(theta, L): #funkcija energije sustava
        energy = 0.0
        for i in range(1, L+1):
            for j in range(1, L+1):
                right = theta[i+1, j] if i < L else theta[1, j] #rubni uvjeti
                up = theta[i, j+1] if j < L else theta[i, 1]
                delta1 = theta[i, j]-right
                delta2 = theta[i, j]-up
                energy += -J*(np.cos(delta1)+np.cos(delta2))
        return energy

def M(theta, L): #funkcija magnetizacije sustava
    mx = np.sum(np.cos(theta[1:L+1, 1:L+1]))
    my = np.sum(np.sin(theta[1:L+1, 1:L+1]))
    return np.sqrt(mx**2 + my**2)

def M_angle(theta, L): #kut magnetizacije u xy-ravnini
    mx = np.sum(np.cos(theta[1:L+1, 1:L+1]))
    my = np.sum(np.sin(theta[1:L+1, 1:L+1]))
    return np.arctan2(my, mx)

def theta2(theta, phi, L): #⟨θ^2⟩
    delta = theta[1:L+1, 1:L+1]-phi
    delta_wrapp = np.angle(np.exp(1j*delta))  #dobivanje kuta u intervalu [-π, π]
    return np.mean(delta_wrapp**2)

for iL in Ls: #petlja po velicinama resetke
    Theta = 2*np.pi*rng.random((iL+2, iL+2)) #matrica kuteva theta [0, 2π)

    dmax = 1.0
    max_dE = 20.0
    w_size = int(max_dE*1000)+1
    w = np.exp(-beta*np.linspace(0.0, max_dE, w_size))
    '''
    Kut theta je kontinuirana varijabla pa je i Boltzmannov faktor promjene energije sustava kontinuirana
    varijabla. Kreira se lista w, gdje je w[j] element ekvivalentan Boltzmannovom faktoru za promjenu
    energije sustava spinova, a j je cijeli broj jednak 1000*dE. Ovo daje rezoluciju energije od 0.001.
    Vrijednost max_dE predstavlja gornju granicu energije za sustav spinova |S(x,y)|=1 gdje se razmatra
    interakcija prvih susjeda, koja ne bi smjera premasiti E=16J, J=1.
    '''
    N = iL**2 #broj cestica u sustavu

    E0 = E(Theta, iL)
    accept = 0.0
    sum_theta2 = 0.0

    for ib in range(1, Nb_skip+Nb+1): #glavna petlja Metropolisovog algoritma
        for ik in range(Nk*N):
            i = rng.integers(1, iL+1)
            j = rng.integers(1, iL+1)

            Theta_old = Theta[i, j]
            dTheta = (rng.random()-0.5)*2*dmax
            Theta_new = (Theta_old+dTheta) % (2*np.pi)

            neighbors = [(i+1 if i < iL else 1, j),
                         (i-1 if i > 1 else iL, j),
                         (i, j+1 if j < iL else 1),
                         (i, j-1 if j > 1 else iL)] #rubni uvjeti

            dE = 0.0
            for x, y in neighbors:
                dE -= J*(np.cos(Theta_new-Theta[x, y])-np.cos(Theta_old-Theta[x, y]))

            if dE <= 0.0:
                Theta[i, j] = Theta_new
                E0 += dE
                accept += 1.0
            else:
                j_idx = int(1000*dE) #indeks j liste w[j]
                if j_idx < w_size and rng.random() < w[j_idx]:
                    Theta[i, j] = Theta_new
                    E0 += dE
                    accept += 1.0

        if ib > Nb_skip:
            phiM = M_angle(Theta, iL) #kut magnetizacije
            sum_theta2 += theta2(Theta, phiM, iL)

        if ib % 100 == 0:
            accept_ib = accept/(Nk*N*ib)
            if accept_ib > acceptance:
                dmax *= 1.05
            elif accept_ib < acceptance:
                dmax *= 0.95

    mean_theta2 = sum_theta2/Nb
    lnN = np.log(N)
    fTheta2.write(f"{iL:>4d} {N:>6d} {mean_theta2:>15.9f} {lnN:>12.7f}\n")

fTheta2.close()

File: seminar_Stoh/sem_b.c
// sem_b.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define PI 3.14159265358979323846
#define MAX_L 230
#define MAX_W 20001

#include "ran1.c"  // Generator slucajnih brojeva

// Alokacija i oslobadjanje 2D matrice
double** alloc_matrix(int n, int m) {
    double** mat = (double**) malloc(n * sizeof(double*));
    for (int i = 0; i < n; i++)
        mat[i] = (double*) malloc(m * sizeof(double));
    return mat;
}

void free_matrix(double** mat, int n) {
    for (int i = 0; i < n; i++)
        free(mat[i]);
    free(mat);
}

int pbc(int x, int L) {
    return (x == 0) ? L : (x == L + 1 ? 1 : x);
}

double E(double** theta, int L, double J) {
    double energy = 0.0;
    for (int i = 1; i <= L; i++) {
        for (int j = 1; j <= L; j++) {
            double delta1 = theta[i][j] - theta[pbc(i+1,L)][j];
            double delta2 = theta[i][j] - theta[i][pbc(j+1,L)];
            energy += -J * (cos(delta1) + cos(delta2));
        }
    }
    return energy;
}

double M_angle(double** theta, int L) {
    double mx = 0.0, my = 0.0;
    for (int i = 1; i <= L; i++) {
        for (int j = 1; j <= L; j++) {
            mx += cos(theta[i][j]);
            my += sin(theta[i][j]);
        }
    }
    return atan2(my, mx);
}

double theta2(double** theta, double phi, int L) {
    double sum = 0.0;
    for (int i = 1; i <= L; i++) {
        for (int j = 1; j <= L; j++) {
            double delta = theta[i][j] - phi;
            double wrapped = atan2(sin(delta), cos(delta));
            sum += wrapped * wrapped;
        }
    }
    return sum / (L * L);
}

int main() {
    double kT = 0.1, beta = 1.0 / kT, J = 1.0;
    int Nk = 4, Nb_skip = 1500, Nb = 10000;
    double acceptance = 0.4;
    double dmax_init = 1.0;
    int Ls[] = {4, 8, 16, 32, 48, 64, 80, 96, 128, 160, 192, 224};
    int num_L = sizeof(Ls) / sizeof(Ls[0]);

    long idum = -1208; // Seed za ran1

    FILE* fTheta2 = fopen("f_theta2_vs_L.dat", "w");
    fprintf(fTheta2, "# L  N  <theta^2>  ln(N)\n");

    for (int idx = 0; idx < num_L; idx++) {
        int L = Ls[idx];
        int N = L * L;
        double dmax = dmax_init;

        double** Theta = alloc_matrix(L+2, L+2);
        for (int i = 0; i <= L+1; i++)
            for (int j = 0; j <= L+1; j++)
                Theta[i][j] = ran1(&idum) * 2 * PI;

        double max_dE = 20.0;
        int w_size = (int)(max_dE * 1000) + 1;
        double* w = (double*) malloc(w_size * sizeof(double));
        for (int j = 0; j < w_size; j++)
            w[j] = exp(-beta * (j / 1000.0));

        double E0 = E(Theta, L, J);
        double accept = 0.0, sum_theta2 = 0.0;

        for (int ib = 1; ib <= Nb_skip + Nb; ib++) {
            for (int ik = 0; ik < Nk * N; ik++) {
                int i = 1 + (int)(ran1(&idum) * L);
                int j = 1 + (int)(ran1(&idum) * L);
                double old_theta = Theta[i][j];
                double dtheta = (ran1(&idum) - 0.5) * 2 * dmax;
                double new_theta = fmod(old_theta + dtheta + 2*PI, 2 * PI);

                double dE = 0.0;
                int ni[4] = {pbc(i+1,L), pbc(i-1,L), i, i};
                int nj[4] = {j, j, pbc(j+1,L), pbc(j-1,L)};

                for (int n = 0; n < 4; n++) {
                    double old_diff = old_theta - Theta[ni[n]][nj[n]];
                    double new_diff = new_theta - Theta[ni[n]][nj[n]];
                    dE -= J * (cos(new_diff) - cos(old_diff));
                }

                if (dE <= 0.0 || ((int)(1000*dE) < w_size && ran1(&idum) < w[(int)(1000*dE)])) {
                    Theta[i][j] = new_theta;
                    E0 += dE;
                    accept += 1.0;
                }
            }

            if (ib > Nb_skip) {
                double phiM = M_angle(Theta, L);
                sum_theta2 += theta2(Theta, phiM, L);
            }

            if (ib % 100 == 0) {
                double accept_ib = accept / (Nk * N * ib);
                if (accept_ib > acceptance) dmax *= 1.05;
                else if (accept_ib < acceptance) dmax *= 0.95;
            }
        }

        double mean_theta2 = sum_theta2 / Nb;
        double lnN = log((double) N);
        fprintf(fTheta2, "%4d %6d %15.9f %12.7f\n", L, N, mean_theta2, lnN);

        free(w);
        free_matrix(Theta, L+2);
    }

    fclose(fTheta2);
    return 0;
}
